### PR TITLE
DeepSeek V4: Support context parallelism with fused MoE (non-DeepEP) 

### PR DESCRIPTION
**Pull request by xu-yfei**
## Motivation

  Currently, Context Parallelism (CP) in DeepSeek-V4 only works with DeepEP as the MoE all-to-all backend. Fused MoE, Marlin MoE, and other MoE backends are not supported under CP, which is a significant limitation for H20-3e            
  deployments:

  - TP8 Fused MoE outperforms DeepEP on H20-3e by a considerable margin.                                                                                                                                                                     
  - DeepEP CUDA graph consumes ~20GB CUDA memory (root cause unknown), severely limiting usable memory for KV cache and batch size.
  - DeepSeek-V4 Pro is constrained: TP16 and PP2 configurations with DeepEP both suffer from the ~20GB CUDA graph memory overhead, leaving insufficient memory for practical workloads.                                                      

## Modifications

1. **`python/sglang/srt/layers/communicator_nsa_cp.py`**:
   - Added `nsa_cp_gather_hidden_states()`: all-gathers hidden states across CP ranks so each rank has the full sequence before MoE (asserts `attn_dp_size == 1` and `attn_tp_size == 1` to verify CP-only mode).
   - Added `nsa_cp_reduce_scatter_hidden_states()`: splits the full hidden states back by CP rank and reduce-scatters, restoring the per-rank token partition after MoE.
   - Refactored `NSACPCommunicateWithAllReduceAndLayerNormFn` and `NSACPCommunicateSummableTensorPairFn` to call these new helpers, removing inline code duplication.

2. **`python/sglang/srt/models/deepseek_v4.py`**:
   - In `DeepseekV4DecoderLayer.forward`: when CP is enabled (`_use_cp`), branches on the MoE backend. With DeepEP, the existing per-rank `input_ids[cp_rank::cp_size]` split is used. With fused MoE (`is_none()`), it calls `nsa_cp_gather_hidden_states()` before MoE and `nsa_cp_reduce_scatter_hidden_states()` after, and passes `use_reduce_scatter=True` to the fused MoE runner.
   - In `DeepseekV4Model.forward`: when using prefill CP with fused MoE, reshapes `input_ids` from per-rank split back to the full-batch layout so the gather step produces correct token indices.

3. **Triton kernel tuning configs**: Added 6 JSON configuration files for the fused MoE runner (`fp8_w8a8`, `block_shape=[128, 128]`) covering `E=256,N=256`, `E=256,N=512`, and `E=384,N=384` for both forward and forward-down passes on NVIDIA H20-3e.

  
## Accuracy Tests

Combined with PR #24704

H20-3e*2，CP8PP2
```
python -m sglang.launch_server \
  --model-path /home/models/DeepSeek-V4-Pro-FP8 \
  --host 0.0.0.0 --port 8188 \
  --trust-remote-code --enable-cache-report --log-level info --enable-metrics \
  --page-size 64 \
  --cuda-graph-max-bs 64 --max-running-requests 64 \
  --mem-fraction-static 0.95 \
  --tp-size 8 --pp-size 2 --nnodes 2 \
  --node-rank {node_rank} --dist-init-addr {master_ip}:25000 \
  --disable-radix-cache \
  --enable-nsa-prefill-context-parallel --nsa-prefill-cp-mode round-robin-split \
  --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4 \
  --thinking-trigger thinking
```

```
python -m sglang.test.run_eval --eval-name gsm8k --host localhost --port 8188 --num-shots 8 --gsm8k-data-path datasets/gsm8k/test.jsonl --num-threads 64 --repeat 1
python -m sglang.test.run_eval --eval-name mmlu --host localhost --port 8188 --num-threads 64 --repeat 1
python -m sglang.test.run_eval --eval-name gpqa --host localhost --port 8188 --num-examples 198 --max-tokens 128000 --repeat 1 --top-p 0.95 --temperature 1.0 --thinking-mode deepseek-v3 --num-threads 64
```

```
GSM8K (8-shot): 96.95%
MMLU:  90.89%
GPQA: 88.9

```

## Speed Tests and Profiling

| Input Len | TTFT (ms) | 
|-----------|-----------|
| 16384 | 1225.91|
| 32768 | 2094.66 |
| 65536 | 3908.72|
| 131072 | 7783.19 |
| 262144 | 16420.96 |
| 524288 | 37940.82 |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26675966144](https://github.com/sgl-project/sglang/actions/runs/26675966144)
Latest PR Test (Extra): :x: [Run #26675966093](https://github.com/sgl-project/sglang/actions/runs/26675966093)

### CODE REVIEW

**Comment by whybeyoung** [comment]
@Fridge003  could you take a review ?

File: python/sglang/srt/models/deepseek_v4.py [hunk 1]
```diff
@@ -983,8 +991,11 @@ def forward(
             forward_batch,
             input_ids=input_ids,
             input_ids_global=input_ids_global,
+            use_reduce_scatter=_use_cp,
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_v4.py:1227]: Looks like `use_reduce_scatter` is useless

File: python/sglang/srt/layers/communicator_dsa_cp.py [hunk 2]
```diff
@@ -47,6 +51,30 @@ def dsa_enable_prefill_cp():
     return is_dsa_enable_prefill_cp()
 
 
+def dsa_cp_gather_hidden_states(hidden_states: torch.Tensor):
+    attn_dp_size = get_attention_dp_size()
+    attn_tp_size = get_attention_tp_size()
+    assert attn_dp_size == 1 and attn_tp_size == 1
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator_dsa_cp.py:58]: Why assert attn_tp_size==1. Will this affect usage of dsa models like glm5

File: python/sglang/srt/layers/communicator_dsa_cp.py [hunk 3]
```diff
@@ -47,6 +51,30 @@ def dsa_enable_prefill_cp():
...
+        hidden_states,
+    )
+    attn_cp_all_gather_into_tensor(hidden_states, local_hidden_states)
+    return hidden_states
+
+
+def dsa_cp_reduce_scatter_hidden_states(hidden_states: torch.Tensor):
+    attn_dp_size = get_attention_dp_size()
+    attn_tp_size = get_attention_tp_size()
+    assert attn_dp_size == 1 and attn_tp_size == 1
```
Comment (Fridge003) [inline python/sglang/srt/layers/communicator_dsa_cp.py:70]: The same

File: python/sglang/srt/models/deepseek_v4.py [hunk 4]
```diff
@@ -957,14 +962,17 @@ def forward(
...
-            assert get_moe_a2a_backend().is_deepep(), (
-                "CP requires DeepEP (moe_a2a_backend == deepep). "
-                "Only DeepEP is tested with CP's per-rank token split."
-            )
-            cp_rank = get_attention_cp_rank()
-            cp_size = get_attention_cp_size()
-            input_ids = input_ids[cp_rank::cp_size].contiguous()
-            input_ids_global = input_ids
+            if get_moe_a2a_backend().is_none():
+                hidden_states = dsa_cp_gather_hidden_states(hidden_states)
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_v4.py:1203]: Can we manage communication in `communicator_dsa_cp.py`. We shouldn't explicitly call communication operators here

File: python/sglang/srt/models/deepseek_v4.py [hunk 5]
```diff
@@ -957,14 +962,17 @@ def forward(
...
+            else:
+                assert get_moe_a2a_backend().is_deepep(), (
+                    "CP requires DeepEP (moe_a2a_backend == deepep). "
+                    "Only DeepEP is tested with CP's per-rank token split."
+                )
+                cp_rank = get_attention_cp_rank()
+                cp_size = get_attention_cp_size()
+                input_ids = input_ids[cp_rank::cp_size].contiguous()
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_v4.py]: I guess line 974 and 975 do the same thing as `cp_round_robin_input_ids`, can we remove the `cp_round_robin_input_ids` function, and reuse these two lines. Since `cp_round_robin_input_ids` is only useful for deepseek v4

**Comment by Fridge003** [comment]
Please add a subtest for your usecase in `test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py`

File: python/sglang/srt/layers/communicator_dsa_cp.py (hunk 2 above)
Comment (xu-yfei) [reply python/sglang/srt/layers/communicator_dsa_cp.py:58]: in server_args.py:
```
                            assert (
                                self.dp_size == 1
                            ), "For round-robin split mode, dp attention is not supported."
                        assert (
                            self.tp_size <= 8
                        ), "Context parallel only supports single machine (tp_size <= 8). Cross-machine CP has precision issues."
                        self.attn_cp_size = self.tp_size // self.dp_size
```
The formula is: `attn_dp_size * attn_tp_size * attn_cp_size = tp_size`.
Therefore, currently only `attn_dp_size` and `attn_tp_size` being set to 1 are supported.

This logic is only executed in non-DeepEP scenarios.
When `attn_dp_size > 1` under non-DeepEP mode, partial gather is required for tokens of each data parallel group.
When `attn_tp_size > 1`, allreduce needs to be performed along the head dimension within tensor parallel ranks first, followed by all-gather to collect tokens across context parallel groups based on `attn_cp_size`. None of these functions are currently supported.

File: python/sglang/srt/models/deepseek_v4.py (hunk 4 above)
Comment (xu-yfei) [reply python/sglang/srt/models/deepseek_v4.py:1203]: Integrating with communicator_dsa_cp may introduce refactoring that involves MLP and attention. We can consider doing the integration refactoring later.

File: python/sglang/srt/models/deepseek_v4.py (hunk 5 above)
Comment (xu-yfei) [reply python/sglang/srt/models/deepseek_v4.py]: This is different. 

In non-deepep scenarios, the inputs to MOE and HashTopK are complete and also follow `cp_round_robin` (i.e., 0,8,16,...,1,9,17,...,2,10,...), so input_ids need to be `cp_round_robin_input_ids`. 

In deepep scenarios, the input to moe is cp-split, i.e., `rank 0: 0,8,16,...`; `rank 1: 1,9,17,...`; and so on. The `input_ids` are also split, directly as `input_ids[cp_rank::cp_size].`

File: python/sglang/srt/models/deepseek_v4.py (hunk 1 above)
Comment (xu-yfei) [reply python/sglang/srt/models/deepseek_v4.py:1227]: It is necessary to set `use_reduce_scatter=True` to skip the all-reduce inside the MLP, and use reduce_scatter instead externally.

**Comment by xu-yfei** [comment]
> Please add a subtest for your usecase in `test/registered/cp/test_deepseek_v4_flash_fp4_b200_cp.py`

done:
```
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_ascii_ratio PASSED [ 56%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_basic_math PASSED [ 62%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_capital_france PASSED [ 68%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_color_completion PASSED [ 75%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_determinism_temp_zero PASSED [ 81%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_gsm8k PASSED [ 87%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_max_token_one PASSED [ 93%]
test_deepseek_v4_flash_fp4_b200_cp.py::TestDSV4FlashFP4B200Balanced_CP_FusedMoe::test_no_repetition_blowup PASSED [100%]
```

File: python/sglang/srt/models/deepseek_v4.py (hunk 5 above)
Comment (xu-yfei) [reply python/sglang/srt/models/deepseek_v4.py]: Or, we can update this code as:
```shell
--- a/python/sglang/srt/models/deepseek_v4.py
+++ b/python/sglang/srt/models/deepseek_v4.py
@@ -1224,10 +1224,6 @@ class DeepseekV4DecoderLayer(nn.Module):
                     "CP requires DeepEP (moe_a2a_backend == deepep). "
                     "Only DeepEP is tested with CP's per-rank token split."
                 )
-                cp_rank = get_attention_cp_rank()
-                cp_size = get_attention_cp_size()
-                input_ids = input_ids[cp_rank::cp_size].contiguous()
-                input_ids_global = input_ids
         elif _use_tp_moe_gather:
             hidden_states, local_hidden_states = (
                 get_global_dp_buffer(get_tp_group()),
@@ -1413,9 +1409,12 @@ class DeepseekV4Model(nn.Module):
             if self.pp_group.is_first_rank:
                 hidden_states = cp_split_and_rebuild_data(forward_batch, hidden_states)
             positions = cp_split_and_rebuild_position(forward_batch, positions)
-            if get_moe_a2a_backend().is_none():
-                input_ids = cp_round_robin_input_ids(input_ids)
-                input_ids_global = input_ids
+            input_ids = cp_round_robin_input_ids(input_ids)
+            if not get_moe_a2a_backend().is_none():
+                cp_rank = get_attention_cp_rank()
+                input_ids = input_ids.tensor_split(self.cp_size)[cp_rank]
+            input_ids_global = input_ids
+
 
         # Reset Compressor's per-step freqs_cis cache from any previous step.
         for _attr in ("freqs_cis_c4", "freqs_cis_c128"):

```

**Review by Fridge003 — APPROVED** [review]
